File: src/main/java/mvu/support/Dispatcher.java
package mvu.support;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class Dispatcher {
	
	final List<Consumer<Action>> parentDispatchers;
	
	final Consumer<Action> dispatcher;
	
	public Dispatcher(Consumer<Action> dispatcher){
		this(new ArrayList<>(), dispatcher);
	}
	
	public Dispatcher(List<Consumer<Action>> parentDispatchers, Consumer<Action> dispatcher){
		this.parentDispatchers = parentDispatchers;
		this.dispatcher = dispatcher;
	}

	public static Dispatcher empty() {
		return new Dispatcher(new ArrayList<>(), action -> {});
	}

	public List<Consumer<Action>> getParentDispatchers() {
		return parentDispatchers;
	}

	public Consumer<Action> getDispatcher() {
		return dispatcher;
	}

	public List<Consumer<Action>> getAllDispatchers(){
		List<Consumer<Action>> allDispatchers = new ArrayList<>();
		allDispatchers.add(dispatcher);
		allDispatchers.addAll(parentDispatchers);
		return allDispatchers;
	}
}

File: src/main/kotlin/mvu/support/Dispatcher.kt
package mvu.support

import java.util.ArrayList

class Dispatcher(private val parentDispatchers: List<(Action) -> Unit>, val dispatcher: (Action) -> Unit) {

    val allDispatchers: List<(Action) -> Unit>
        get() {
            val allDispatchers = ArrayList<(Action) -> Unit>()
            allDispatchers.add(dispatcher)
            allDispatchers.addAll(parentDispatchers)
            return allDispatchers
        }

    constructor(dispatcher: (Action) -> Unit) : this(ArrayList<(Action) -> Unit>(), dispatcher)

}

fun emptyDispatcher(): Dispatcher {
    return Dispatcher(ArrayList()) { }
}
